compiler/hash-parser/src/lexer.rs

```diff
@@ -0,0 +1,618 @@
+                        TokenErrorKind::MalformedNumericalLiteral,
+                        Location::span(start, self.offset.get()),
+                    ));
+                }
+
+                let exp = self.eat_exponent()?;
+
+                // if an exponent was speified, as in it is non-zero, we need to apply the exponent to
+                // the float literal.
+                let value = if exp != 0 { num.unwrap() * 10f64.powi(exp) } else { num.unwrap() };
+
+                Ok(TokenKind::FloatLiteral(value))
+            }
+            _ => {
+                let num = pre_digits.parse::<u64>();
+
+                if num.is_err() {
+                    return Err(TokenError::new(
+                        Some("Integer literal too large.".to_string()),
+                        TokenErrorKind::MalformedNumericalLiteral,
+                        Location::span(start, self.offset.get()),
+                    ));
+                }
+
+                Ok(TokenKind::IntLiteral(num.unwrap()))
+            }
+        }
+    }
+
+    /// Consume an exponent for a float literal
+    fn eat_exponent(&self) -> TokenResult<i32> {
+        debug_assert!(matches!(self.prev.get(), 'e' | 'E'));
+        let start = self.offset.get();
+
+        // Check if there is a sign before the digits start in the exponent...
+        let negated = if self.peek() == '-' {
+            self.next();
+            true
+        } else {
+            false
+        };
+
+        let num = self.eat_decimal_digits().collect::<String>().parse::<i32>();
+
+        // Ensure that the digit parsing was ok
+        if num.is_err() {
+            return Err(TokenError::new(
+                Some("Invalid float exponent.".to_string()),
+                TokenErrorKind::MalformedNumericalLiteral,
+                Location::span(start, self.offset.get()),
+            ));
+        }
+
+        let num = num.unwrap();
+
+        if negated {
+            Ok(-num)
+        } else {
+            Ok(num)
+        }
+    }
+
+    /// Consume only decimal digits up to encountering a non-decimal digit
+    /// whilst taking into account that the language supports '_' as digit
+    /// separators which should just be skipped over...
+    pub(crate) fn eat_decimal_digits(&self) -> impl Iterator<Item = char> + '_ {
+        self.eat_while(|peeked| matches!(peeked, '0'..='9' | '_'))
+            .filter(|e| *e != '_')
+    }
+
+    /// Transform an ordinary character into a well known escape sequence specified by the
+    /// escape literal rules. More information about the escape sequences can be found at
+    /// [escape sequences](https://hash-org.github.io/lang/basics/intro.html)
+    fn char_from_escape_seq(&self) -> TokenResult<char> {
+        debug_assert!(self.prev.get() == '\\');
+
+        // @@Incomplete: come up with a better algorithm to transform escaped literals, rather than manual
+        // transformations!
```

Comment:
What do you have in mind?

Reply:
Old comment before the method was implemented.